include/deal.II/base/std_cxx11/unique_ptr.h

```diff
@@ -33,4 +33,70 @@ DEAL_II_NAMESPACE_OPEN
 namespace std_cxx1x = std_cxx11;
 DEAL_II_NAMESPACE_CLOSE
 
+
+DEAL_II_NAMESPACE_OPEN
+
+/**
+ * Convert an object of type `std::unique_ptr<From>` to an object of
+ * type `std::unique_ptr<To>`, where it is assumed that we can cast
+ * the pointer to `From` to a pointer to `To` using a `dynamic_cast`
+ * -- in other words, we assume that `From` and `To` are connected
+ * through a class hierarchy, and that the object pointed to is in
+ * fact of a type that contains both a `From` and a `To`. An example
+ * is if either `To` is derived from `From` or the other way around.
+ *
+ * The function throws an exception of type `std::bad_cast` if the
+ * `dynamic_cast` does not succeed. This is the same exception you
+ * would get if a regular `dynamic_cast` between object types (but not
+ * pointer types) does not succeed.
+ *
+ * An example of how this function works is as follows:
+ * @code
+ *   class B { ... };            // A base class. Assume that it has virtual
+ *                               // functions so that dynamic_cast can work.
+ *   class D : public B { ... }; // A derived class
+ *
+ *
+ *   std::unique_ptr<B> create_object (...) {...}  // A factory function
+ *
+ *   void foo (...)
+ *   {
+ *     std::unique_ptr<B> b = create_object (...);
+ *
+ *     // Assume that we know for some reason that the object above must
+ *     // have created a D object but returned it as a std::unique_ptr<B>.
+ *     // In order to access the D functionality, we need to cast the
+ *     // pointer. Use the equivalent to dynamic_cast:
+ *     std::unique_ptr<D> d = dynamic_unique_cast<D>(std::move(b));
+ *
+ *     // If the object really was a D, then 'd' now points to it. Note
+ *     // also that in accordance with the semantics of std::unique_ptr,
+ *     // it was necessary to std::move the 'b' object, and indeed 'b'
+ *     // now no longer pointers to anything -- ownership has been
+ *     // transferred to 'd'!
+ * @endcode
+ *
+ * @note This function does not try to convert the `Deleter` objects stored
+ *   by `std::unique_ptr` objects. The function therefore only works if the
+ *   deleter objects are at their defaults.
+ */
```

Comment:
Can we have a `static_assert` for this? Something like
```
static_assert(std::is_same<std::unique_ptr<From>::deleter_type, std::default_delete<From>>::value, 
              "The std::unique_ptr given must have a default deleter!");
```
should work.

Reply:
We don't need this -- the argument of the function is a `std::unique_ptr<From>`, which implies that the second argument is at its default value. We *know* that whenever the function is called the static assertion would be satisfied.

Reply:
Maybe to be more clear: The compiler will reject a call to
```
  std::unique_ptr<B,MyDeleter> b;
  std::unique_ptr<D,SomeDeleter> y = dynamic_unique_cast<D>(std::move(b));
```
because the argument `b` does not match the type of the argument of `dynamic_unique_cast`.

Reply:
Thanks for clarifying! Indeed, 
```
<source>: In function 'void test()':
<source>:39:62: error: no matching function for call to 'dynamic_unique_cast<B>(std::remove_reference<std::unique_ptr<D, void (*)(D*)>&>::type)'
    std::unique_ptr<B> b = dynamic_unique_cast<B>(std::move(d));
                                                              ^
<source>:5:2: note: candidate: template<class To, class From> std::unique_ptr<_Tp> dynamic_unique_cast(std::unique_ptr<From>&&)
  dynamic_unique_cast (std::unique_ptr<From/*, Deleter*/> &&p)
  ^~~~~~~~~~~~~~~~~~~
<source>:5:2: note:   template argument deduction/substitution failed:
<source>:39:62: note:   mismatched types 'std::default_delete<From>' and 'void (*)(D*)'
    std::unique_ptr<B> b = dynamic_unique_cast<B>(std::move(d));
```
should be quite understandable.

Reply:
Maybe as an additional data point: deleters are generally only useful for objects that have no destructors. Say, pointers to PETSc objects that need to be destroyed by hand via calling a special function. On the other hand, if you're using `dynamic_cast`, then that can only work if you have a class hierarchy, and these classes furthermore have virtual members. Chances are that you will have a virtual destructor (and if you don't, you should!), which implies that the default deleter is doing the right thing in calling this destructor when destroying the object.

In other words, for classes for which you want to use this function, you will rarely need anything other than the default deleter.

Reply:
It would be nice (but it is certainly not a blocker) to mention `std::default_delete` by name here for clarity.